gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/ExpireContentsProcedure.java

```diff
@@ -0,0 +1,292 @@
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow internalRow) {
+    String gcCatalogName = internalRow.getString(0);
+    String gcOutputBranchName = internalRow.getString(1);
+    String gcOutputTableIdentifier = internalRow.getString(2);
+    Map<String, String> nessieClientConfig = new HashMap<>();
+    MapData map = internalRow.getMap(3);
+    for (int i = 0; i < map.numElements(); i++) {
+      nessieClientConfig.put(
+          map.keyArray().getUTF8String(i).toString(), map.valueArray().getUTF8String(i).toString());
+    }
+    String runId = !internalRow.isNullAt(4) ? internalRow.getString(4) : null;
+    boolean dryRun = !internalRow.isNullAt(5) && internalRow.getBoolean(5);
+
+    IdentifiedResultsRepo identifiedResultsRepo =
+        new IdentifiedResultsRepo(
+            spark(), gcCatalogName, gcOutputBranchName, gcOutputTableIdentifier);
+
+    if (runId == null) {
+      runId = getLatestCompletedRunID(gcOutputTableIdentifier, identifiedResultsRepo);
+    }
+
+    try (FileIO fileIO = getFileIO(nessieClientConfig)) {
+      Dataset<Row> expiredContentsDF = getExpiredContents(runId, identifiedResultsRepo, fileIO);
+      if (!dryRun) {
+        expiredContentsDF =
+            expiredContentsDF.map(
+                new DeleteFunction(fileIO), RowEncoder.apply(expiredContentsDF.schema()));
+      }
+      List<Row> rows = expiredContentsDF.collectAsList();
+      return rows.stream()
+          .map(
+              row ->
+                  GCProcedureUtil.internalRow(
+                      // "content_id", "deleted_files_type", "deleted_files_count",
+                      // "deleted_files_list"
+                      row.getString(0), row.getString(1), row.getInt(3), row.getList(2)))
+          .toArray(InternalRow[]::new);
+    }
+  }
+
+  private FileIO getFileIO(Map<String, String> nessieClientConfig) {
+    Configuration config = spark().sparkContext().hadoopConfiguration();
+    String fileIOImpl = nessieClientConfig.get(CatalogProperties.FILE_IO_IMPL);
+    return fileIOImpl == null
+        ? new HadoopFileIO(config)
+        : CatalogUtil.loadFileIO(fileIOImpl, nessieClientConfig, config);
+  }
+
+  private static Dataset<Row> getExpiredContents(
+      String runId, IdentifiedResultsRepo identifiedResultsRepo, FileIO fileIO) {
+    // Read the expired content-rows from output table for this run id
+    Dataset<Row> expiredContents = identifiedResultsRepo.collectExpiredContentsAsDataSet(runId);
+    Dataset<Row> expiredContentsDF = computeAllFiles(fileIO, expiredContents);
+    // Read the live content-rows from output table for this run id
+    Dataset<Row> liveContents = identifiedResultsRepo.collectLiveContentsAsDataSet(runId);
+    Dataset<Row> liveContentsDF = computeAllFiles(fileIO, liveContents);
+    // remove the files which are used by live contents
+    Dataset<Row> expiredFilesDF = expiredContentsDF.except(liveContentsDF);
+    // final output
+    // Example output row:
```

Comment:
If I understand correctly, `getExpiredContents` returns rows with _collections_ of files to delete.
Those collections can become very very big (expiring many 1000s of files at once) - and result in too much heap pressure - resulting in out-of-memory-errors.
IIRC Ryan's left a comment about not using collections in `Row`s a while ago as well.